### PR TITLE
#views users for created dashboards on profile page

### PR DESCRIPTION
**Pull request by vera-liu**
I'm not sure if this is the standard way to add index. Any suggestions are appreciated.

Done:
 - Added #views as column to dashboards and slices
 - Pre-populates #views from logs table 
 - Include # views in CreatedContent component

Note:
 - for postgres users, #views started counting from the time of this PR

@mistercrunch @bkyryliuk 

screen shot 2016-12-15 at 5 21 06 pm

### CODE REVIEW

**Comment by ascott** [comment]
should we add views and users to slices as well? if not i think we should align the `favorited` columns so the line up vertically.

**Comment by ascott** [comment]
not too familiar with sqlalchemy indexes but otherwise LGTM. @mistercrunch @bkyryliuk any comments on the indexing code here?

File: superset/views.py [hunk 1]
```diff
@@ -1801,26 +1801,38 @@ def fave_dashboards(self, user_id):
...
     def created_dashboards(self, user_id):
         Dash = models.Dashboard  # noqa
+        Log = models.Log
         qry = (
             db.session.query(
                 Dash,
+                sqla.func.count(),
+                sqla.func.count(Log.user_id.distinct()),
             )
+            .join(Log)
```
Comment (mistercrunch) [inline superset/views.py]: `outerjoin`

also note that the query may perform better as a join to a subquery on the Log table that aggregates by dashboard id

File: superset/views.py [hunk 2]
```diff
@@ -1801,26 +1801,38 @@ def fave_dashboards(self, user_id):
...
             )
+            .filter(
+                sqla.and_(
+                    Log.dashboard_id == Dash.id,
+                )
+            )
             .order_by(
                 Dash.changed_on.desc()
             )
+            .group_by(Dash.id)
```
Comment (mistercrunch) [inline superset/views.py]: I'm not sure but it seems like this should be `.group_by(Dash)` or individual fields should be specified in the `select` and `group_by` clauses

File: superset/models.py [hunk 3]
```diff
@@ -2219,12 +2218,13 @@ class Log(Model):
...
     user_id = Column(Integer, ForeignKey('ab_user.id'))
-    dashboard_id = Column(Integer)
-    slice_id = Column(Integer)
+    dashboard_id = Column(Integer, ForeignKey('dashboards.id'))
+    slice_id = Column(Integer, ForeignKey('slices.id'))
     json = Column(Text)
     user = relationship('User', backref='logs', foreign_keys=[user_id])
     dttm = Column(DateTime, default=datetime.utcnow)
     dt = Column(Date, default=date.today())
+    views = Index(dashboard_id, user_id, action, dttm)
```
Comment (mistercrunch) [inline superset/models.py]: I think the common way is more something like this:
```
    __table_args__ = (
        Index(foo),
        Index(bar),
    )
```

File: superset/migrations/versions/e27bbff06636_.py [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+# revision identifiers, used by Alembic.
+revision = 'e27bbff06636'
+down_revision = 'e46f2d27a08e'
+
+from alembic import op
+import sqlalchemy as sa
+
+
+def upgrade():
+    op.create_index(
```
Comment (mistercrunch) [inline superset/migrations/versions/e27bbff06636_.py]: just curious, was this auto-generated or did you copy/paste it from some place?

**Review by mistercrunch — COMMENTED** [review]
I'm a bit worried about how this can introduce lag in loading simple dashbaord or slice lists. How does it perform on production with millions of entries in the logs table?

If we're going to refer to views per dashboard commonly I think we should maintain a counter in the dashboard. That would mean a database migration script to add the field and prepolutate it, and and incrementing the view counter on the dashboard endpoint.

File: superset/migrations/versions/e27bbff06636_.py (hunk 4 above)
Comment (vera-liu) [inline superset/migrations/versions/e27bbff06636_.py]: this is auto-generated and modified

**Comment by vera-liu** [comment]
I reverted the original query appraoch, added the views column to dashboards and slices, pre populated them with Log data in the migration file.

**Comment by mistercrunch** [comment]
LGMT

**Pull request merged**